### PR TITLE
Bug 2037209: refactor AlibabaResourceReference

### PR DESCRIPTION
**Pull request by elmiko**
This change updates the AlibabaResourceReference struct to follow a
discriminated union pattern. It now contains a `Type` field which will
indicate whether the resource reference is an ID, a name, or a list of tags. The
`ID` and `Tags` fields have been converted to pointers in following with
that pattern, and some constant values have been added for the various
types.

Additionally, it also converts the previous ResourceGroupID field into
ResourceGroup. This change is being made to convert that field from a
string into the AlibabaResourceReference type so that we can support
multiple types of lookup for the resource group. There are some cases,
notably during installation, where the ID may not be known, and in these
cases the Tags will be used to find the resource group. This change to
the structure gives us a method to detect this option when receiving the
API object in the machine api actuator.

This change makes the `VpcID`, `SecurityGroups`, `VSwitch`, and
`ResourceGroups` fields required byu removing the `+optional` and
`omitempty` annotations. This change is being made to reflect the best
practices when deploying OpenShift on Alibaba Cloud.

When deploying on OpenShift, only the VPC mode is allowed. This means
that the user must specify a VPC ID or the instance creation will fail.
When creating instances in VPC mode, the user must specify a vSwitch or
the operation will fail.

### CODE REVIEW

**Comment by elmiko** [comment]
i suppose we don't necessarily need the hold, but i just want to make sure we don't break the alibaba actuator.

**Comment by JoelSpeed** [comment]
This matches my expectations based on our conversation earlier
/lgtm

File: machine/v1/types_alibabaprovider.go [hunk 1]
```diff
@@ -153,13 +161,16 @@ type AlibabaCloudMachineProviderConfig struct {
...
 type AlibabaResourceReference struct {
+	// type identifies the resource reference type for this entry.
+	Type AlibabaResourceReferenceType `json:"type"`
+
 	// ID of resource
 	// +optional
-	ID string `json:"id,omitempty"`
+	ID *string `json:"id,omitempty"`
 
 	// Tags is a set of metadata based upon ECS object tags used to identify a resource
```
Comment (deads2k) [inline machine/v1/types_alibabaprovider.go:171]: can you clarify what happens if Tags produces multiple matches?

**Comment by deads2k** [comment]
I'd like the comments on the usage of the resource reference clarified and then this change lgtm.

File: machine/v1/types_alibabaprovider.go (hunk 1 above)
Comment (elmiko) [reply machine/v1/types_alibabaprovider.go:171]: will do

**Comment by elmiko** [comment]
@deads2k updated to include more information about the cases when multiple results could be returned from a tag-based search. i also added a note to the resource struct itself, but it felt a little clunky. let me know what you think.

**Comment by elmiko** [comment]
removing the hold as we have a patch ready to go
/hold cancel

**Comment by elmiko** [comment]
it sounds like adding the name field will help with the resource group issue we are tracking
/hold cancel

File: machine/v1/types_alibabaprovider.go [hunk 2]
```diff
@@ -120,9 +134,11 @@ type AlibabaCloudMachineProviderConfig struct {
 	// +optional
 	RAMRoleName string `json:"ramRoleName,omitempty"`
 
-	// ResourceGroupID is the unique ID of the resource group to which to assign the instance.
+	// ResourceGroup references the resource group to which to assign the instance.
+	// A reference holds either the resource group ID, the resource name, or the required tags to search.
```
Comment (JoelSpeed) [inline machine/v1/types_alibabaprovider.go:138]: Resource groups cannot use tags to search, we should update this comment to note that tags can't be used with a resource group lookup

Comment (elmiko) [reply machine/v1/types_alibabaprovider.go:138]: i'll add that

File: machine/v1/types_alibabaprovider.go [hunk 3]
```diff
@@ -111,7 +123,9 @@ type AlibabaCloudMachineProviderConfig struct {
 	// +optional
 	SystemDisk SystemDiskProperties `json:"systemDisk,omitempty"`
 
-	// VSwitch is a reference to the vswitch to use for this instance
+	// VSwitch is a reference to the vswitch to use for this instance.
+	// A reference holds either the vSwitch ID, the resource name, or the required tags to search.
+	// When more than one vSwitch is returned for a tag search, only the first vSwitch returned will be used.
```
Comment (deads2k) [inline machine/v1/types_alibabaprovider.go:128]: is this ordering guaranteed to be consistent from the API you're searching with?

**Comment by deads2k** [comment]
I have a question and I'm curious what happens on zero matches for these.  The change to the reference looks good.

/approve

File: machine/v1/types_alibabaprovider.go (hunk 3 above)
Comment (elmiko) [reply machine/v1/types_alibabaprovider.go:128]: that is a great question, it's not immediately clear from the docs[0] if the ordering is consistent. i will have to confer with alibaba about this. should we add a note here about that?


[0] https://www.alibabacloud.com/help/en/doc-detail/110425.htm?spm=a2c63.p38356.0.0.322619b8Xb6y0V#doc-api-Ecs-ListTagResources

**Comment by elmiko** [comment]
> I have a question and I'm curious what happens on zero matches for these.

if there are zero matches for security group or vswitch, that will produce an error when creating the Machine. @JoelSpeed and i are still working through what should happen if there are zero resource groups found. i had thought it should also cause an error, but Joel is pointing out that perhaps some alibaba api queries are acceptable without a resource group. if that is the case, then a zero result on a resource group search wouldn't, by itself, indicate a failure.

**Comment by JoelSpeed** [comment]
I believe for resource groups, if there are no matches, we would send no resource group ID within the request, in which case, the Alibaba API puts the resource in the default resource group, which may or may not be acceptable.

I think if a user specifies tags/name to match a resource group, and we can't find it, that should be an error. I would only allow this fallback to the default resource group when no resource group configuration is given explicitly

**Comment by elmiko** [comment]
> I think if a user specifies tags/name to match a resource group, and we can't find it, that should be an error. I would only allow this fallback to the default resource group when no resource group configuration is given explicitly

this would be in a situation where the user has omitted the `ResourceGroup` field completely?

that makes me wonder if we should make it a pointer?

**Comment by JoelSpeed** [comment]
> this would be in a situation where the user has omitted the ResourceGroup field completely?
> that makes me wonder if we should make it a pointer?

Correct. We should make sure that it is in fact optional, and if it is, make the field a pointer IMO

**Comment by elmiko** [comment]
i'm putting a hold here just to make sure we don't merge before we figure out this resource group issue
/hold

**Comment by elmiko** [comment]
i think we've resolved all the issue around these fields. i have updated the description, and my last change is essentially the following:

This change makes the `VpcID`, `SecurityGroups`, `VSwitch`, and
`ResourceGroups` fields required byu removing the `+optional` and
`omitempty` annotations. This change is being made to reflect the best
practices when deploying OpenShift on Alibaba Cloud.

When deploying on OpenShift, only the VPC mode is allowed. This means
that the user must specify a VPC ID or the instance creation will fail.
When creating instances in VPC mode, the user must specify a vSwitch or
the operation will fail.

i would like to get @JoelSpeed to review this again and i think we should consult Alibaba as well
/hold cancel

File: machine/v1/types_alibabaprovider.go [hunk 4]
```diff
@@ -111,18 +122,22 @@ type AlibabaCloudMachineProviderConfig struct {
...
 	RAMRoleName string `json:"ramRoleName,omitempty"`
 
-	// ResourceGroupID is the unique ID of the resource group to which to assign the instance.
-	// +optional
-	ResourceGroupID string `json:"resourceGroupId,omitempty"`
+	// ResourceGroup references the resource group to which to assign the instance.
+	// A reference holds either the resource group ID, the resource name, or the required tags to search.
+	// When more than one resource group are returned for a search, an error will be produced and the Machine will not be created.
+	// Resource Groups do not support searching by tags.
+	ResourceGroup AlibabaResourceReference `json:"resourceGroup"`
```
Comment (JoelSpeed) [inline machine/v1/types_alibabaprovider.go:140]: Based on some of the conversations I've seen on slack, I think this should be optional

**Comment by jianli-wei** [comment]
"build openshift/api#1096,openshift/cluster-api-provider-alibaba#26" failed, see https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-gcp-modern/1483070108938014720

**Comment by jianli-wei** [comment]
Anyone could help checking why "build openshift/api#1096,openshift/cluster-api-provider-alibaba#26" cannot succeed? 
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-gcp-modern/1484411008238751744

**Comment by elmiko** [comment]
@jianli-wei i think those errors are unrelated to this pr.

**Comment by elmiko** [comment]
@jianli-wei it looks like you might have launched a cluster from the ci-bot with those 2 PRs ?

assuming that is the case, i think we need to update the installer as well before this will fully work. i am creating a PR to make sure that the installer is updated with these api changes.

**Comment by elmiko** [comment]
this is also needed by https://github.com/openshift/installer/pull/5562

**Comment by JoelSpeed** [comment]
/lgtm
/hold

I believe this is ready to go now, @elmiko feel free to hold cancel if you're not aware of any further changes required here

**Comment by elmiko** [comment]
thanks @JoelSpeed and @deads2k , i am removing the hold as we have the follow up patches ready to go.
/hold cancel

**Pull request merged**